contracts/switchboard/default-switchboards/FastSwitchboard.sol

```diff
@@ -62,41 +72,63 @@ contract FastSwitchboard is SwitchboardBase {
     /**
      * @dev Function to attest a packet
      * @param packetId_ Packet ID
+     * @param proposalId_ Proposal ID
      * @param signature_ Signature of the packet
      */
-    function attest(bytes32 packetId_, bytes calldata signature_) external {
+    function attest(
+        bytes32 packetId_,
+        uint256 proposalId_,
+        bytes calldata signature_
+    ) external {
         uint32 srcChainSlug = uint32(uint256(packetId_) >> 224);
+        bytes32 root = socket__.packetIdRoots(packetId_, proposalId_);
+        if (root == bytes32(0)) revert InvalidRoot();
+        // Should we change the signature to include proposalId as well?
         address watcher = signatureVerifier__.recoverSignerFromDigest(
-            keccak256(abi.encode(address(this), chainSlug, packetId_)),
+            keccak256(
+                abi.encode(address(this), chainSlug, packetId_, proposalId_)
+            ),
             signature_
         );
 
-        if (isAttested[watcher][packetId_]) revert AlreadyAttested();
+        if (isAttested[watcher][root]) revert AlreadyAttested();
         if (!_hasRoleWithSlug(WATCHER_ROLE, srcChainSlug, watcher))
             revert WatcherNotFound();
 
-        isAttested[watcher][packetId_] = true;
-        attestations[packetId_]++;
+        isAttested[watcher][root] = true;
+        attestations[root]++;
 
-        if (attestations[packetId_] >= totalWatchers[srcChainSlug])
-            isPacketValid[packetId_] = true;
+        if (attestations[root] >= totalWatchers[srcChainSlug])
+            isRootValid[root] = true;
```

Comment:
Explain how storing things against root helps us avoid re-attestations in case of duplicate propose